atmega: refactor cpu/board code and build/flash variables

kYc0o opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR proposes a refactor of the atmega based boards to match the code reuse done in other architectures. In summary:

  - Added atmega.inc.mk for toolchain variables
  - Added avrdude.inc.mk for flashing variables
  - Move cpu.c and startup.c to cpu/atmega_common
  - Share cpu_conf.h to unify the stack memory definitions
  - Cleanup all unnecessary code

I hope I factored out as much as possible, if you find something else that can be reused please let me know.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->
Fixes #9001

kYc0o commented:
@Josar I'm not sure if you included arduino-common because you really need it or just to import toolchain and flasher requirements. Can you confirm?

Josar commented on cpu/atmega_common/cpu.c:
```diff
@@ -28,6 +33,17 @@
 #define ENABLE_DEBUG (0)
```
```
#include <avr/pgmspace.h>

#include "cpu.h"
```

missing new line?

aabadie commented on boards/common/arduino-atmega/Makefile.include:
```diff
@@ -3,19 +3,7 @@ include $(RIOTBOARD)/common/arduino-atmega/Makefile.dep
-OFLAGS += -j .text -j .data
-export FFLAGS += -c $(PROGRAMMER) $(PROGRAMMER_FLAGS) -F -D -U flash:w:$(HEXFILE)
+# include flashing tool
+include $(RIOTMAKE)/tools/avrdude.inc.mk
```
missing newline at end of file

aabadie commented on cpu/atmega_common/Makefile.include:
```diff
@@ -1,28 +1,4 @@
-  # See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59396
-  export LINKFLAGS += -Wno-error
-endif
+include $(RIOTMAKE)/arch/atmega.inc.mk
```
missing newline at end of file

Josar commented:
@kYc0o if i remember it correctly it was requested to minimize code duplicates.
Maybe @shr70 can comment on this.
But i think you are right it is just for the toolchain and flasher.
This was still something which i was not very happy with as the jiminy is not related to arduino.

shr70 commented:
@kYc0o @Josar Yes, we only included the arduino-common due to the flasher and build commands. This was requested in the review.

Josar commented:
@kYc0o flashing seems to work.

May i ask you to also change this depency to arduino?
https://github.com/RIOT-OS/RIOT/blob/d6c63592bbde738873c684e49fc1293ab90ddcf1/boards/jiminy-mega256rfr2/Makefile.features#L1

I would suggest just add all `FEATURES_PROVIDED` to the jiminy Makefile.features. Or do i miss something which makes this a bad idea?

kaspar030 commented:
> +161 −791

I like!

ZetaR60 commented on cpu/atmega256rfr2/include/cpu_conf.h:
```diff
@@ -1,51 +0,0 @@
-#define THREAD_STACKSIZE_DEFAULT   (512)
-#endif
-
-#define THREAD_STACKSIZE_IDLE      (129)
```
Note change of behavior: typo fixed.

kYc0o replied:
It wasn't a typo actually. The authors of this port (@Josar and @shr70) claim that with this change `printf()` works from the interrupt context. Can you guys confirm?

Josar replied:
If i remember correctly, I think it has to do with this line.

https://github.com/RIOT-OS/RIOT/blob/af829e2da41982bc903fe5e0149745cfea7f863c/core/include/debug.h#L55

If you use `DEVELHELP` and there are interrupts when in idle thread there will just be the message
```
"Cannot debug, stack too small")
```

kYc0o replied:
I agree with @ZetaR60, and should be done in a folo. I can take care of it when this PR is merged.

ZetaR60 commented on cpu/atmega_common/include/cpu_conf.h:
```diff
@@ -28,25 +31,30 @@
-
 #ifndef THREAD_STACKSIZE_DEFAULT
-#define THREAD_STACKSIZE_DEFAULT   (256)
+#define THREAD_STACKSIZE_DEFAULT   (512)
```
Note change in behavior here. See related: #9001 and #9111

kYc0o commented:
Another update, and I think I'm triggering a more larger discussion.

Currently, atmega based boards are "forced" to import the CPU features from arduino-atmega/Makefile.features, which in my opinion doesn't make sense since 3 atmega based
boards are not arduinos, thus shouldn't have nothing to do with it.

I moved the features to cpu/atmega_common/Makefile.features where I thing they suit better, actually they are implemented in the same tree, which IMHO it's clear to understand why the features are defined there as well.

Finally, I left the arduino feature where I think it belongs, arduino-atmega/Makefile.features.

If nobody objects, I'd propose to start a "cleanup" in the future of these definitions, since currently every single board define features which are implemented at the cpu level.

Additionally, I removed the grouping needed previously by the CI, and if I'm not mistaken, Murdock doesn't need it anymore (@kaspar030 please correct me if I'm wrong).

ZetaR60 commented on cpu/atmega1281/startup.c:
```diff
@@ -1,72 +0,0 @@
- * This way there should be no way to call main directly.
- */
-void init7_ovr(void) __attribute__((naked)) __attribute__((section(".init7")));
-void init8_ovr(void) __attribute__((naked)) __attribute__((section(".init8")));
```
Note: various attributes changed in this file with refactor.

ZetaR60 commented on cpu/atmega328p/startup.c:
```diff
@@ -1,72 +0,0 @@
- * This way there should be no way to call main directly.
- */
-void init7_ovr(void) __attribute__((naked)) __attribute__((section(".init7")));
-void init8_ovr(void) __attribute__((naked)) __attribute__((section(".init8")));
```
Note: various attributes changed in this file with refactor.

kYc0o replied:
If I'm not mistaken, those lines are the equivalent of:

```c
__attribute__((used, naked)) void init7_ovr(void)
{
    __asm__ ("call reset_handler");
}

__attribute__((used, naked)) void init8_ovr(void)
{
    __asm__ ("jmp exit");
}
```

ZetaR60 replied:
I think you're right. Nevermind about this one then.

kaspar030 commented:
> If nobody objects, I'd propose to start a "cleanup" in the future of these definitions, since currently every single board define features which are implemented at the cpu level.

Some features depend on cpu support *and* board configuration (periph_conf.h). That's why e.g., periph_uart is usually in a board's Makefile.features, even though the CPU provides the underlying implementation.

kaspar030 commented:
I think @bergzand is right, this PR is getting very large. Difficult to spot problems.

ZetaR60 commented:
I'm not sure about moving Makefile.features entries to a common area. How do you handle, for instance, a board that has all of the ADC pins connected to LED outputs? That board would not provide ADC support, even though the CPU supports it.

kYc0o commented:
My logic behind the features is that they might be provided by the CPU, but might not be used by the board. I think it's manageable to make a difference between both and express it in the code. Actually, shouldn't my change break some compiles? I was experimenting with the change expecting Murdock to complain about it.

kYc0o commented on makefiles/tools/avrdude.inc.mk:
```diff
@@ -0,0 +1,15 @@
+
+# don't force to flash HEXFILE, but set it as default
+FLASHFILE ?= $(HEXFILE)
+FFLAGS += -c $(PROGRAMMER) $(PROGRAMMER_FLAGS) -F -D -U flash:w:$(HEXFILE)
```
Thanks for the explanation! I'll change this to reflect the differences between arduinos and other atmega based boards.

kYc0o commented on makefiles/arch/atmega.inc.mk:
```diff
@@ -0,0 +1,39 @@
+# Export the peripheral drivers to be linked into the final binary, for now
+# only atmega126rfr2 has periph drivers
+ifeq ($(CPU), atmega256rfr2)
+  USEMODULE += periph
```
I did it this way because is the preferred way to do it in RIOT. See makefiles/arch/cortexm.inc.mk for another example.

ZetaR60 replied:
That case with cortexm seems like a much lower level use than a USEMODULE. My personal preference would be to have it in the cpu specific code, but that's just me.

kYc0o commented on makefiles/arch/atmega.inc.mk:
```diff
@@ -0,0 +1,39 @@
+
+# explicitly tell the linker to link the syscalls and startup code.
+# without this the interrupt vectors will not be linked correctly!
+UNDEF += $(BINDIR)/atmega_common/startup.o
```
Which is strange... shouldn't this be the same on all platforms? I tried both cases and both works... I'm missing something somehow.

kYc0o replied:
As far as I can track, this was introduced by the very first atmega platform supported (atmega2560) bye @N8Fear. Do you know why it was necessary (besides the comment about the vectors)? I'll make a test without and give feedback.

maribu replied:
Anything new on this front? It would be nice to drop the line (if no longer needed) or to extend the comment to be more elaborate on the technical details.

kYc0o commented on cpu/atmega_common/cpu.c:
```diff
@@ -1,33 +1,50 @@
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
+#if !defined (CPU_ATMEGA256RFR2)
```
It's a bit difficult because it would clash with cpu.c from atmega_common. The options are this `#if !defined` or copy all the code in atmega_common to atmega256rfr2. Do you have an idea on how to avoid the clash?

ZetaR60 replied:
What about using something similar to the header guards:

```
#ifndef ATMEGA_CPU_C
#define ATMEGA_CPU_C

/* Code here */

#endif /* ATMEGA_CPU_C */
```

kYc0o replied:
Hmmm either I don't understand what you mean or I don't know how those guards work, but it's not that thing to avoid concurrency in a more "first come, first served" way? I'd need to investigate if the cpu.c from atmega256rfr2 comes first, so then it might work. Though, it's not common to do that in RIOT...

ZetaR60 replied:
I think a better way to do this would be to move all of the specific code into a function that is called by cpu_init. Sort of like how the LED code was refactored for the ATmega-based boards.

kYc0o commented on boards/mega-xplained/Makefile.include:
```diff
@@ -2,23 +2,14 @@
-export PROGRAMMER ?= buspirate
+FFLAGS += -p m1284p
 
-export PROGRAMMER_FLAGS = -P /dev/ttyUSB0
```
I have added an `AVRDUDE_PORT` variable for this case. Will use it.

kYc0o replied:
I refactored the flasher options for avrdude. Can you retest? It should work out of the box for your buspirate adapter.

ZetaR60 replied:
Seems to be working fine now, at least on Linux.

kYc0o commented:
What's next then on this PR? Should I squash to try a split in different PRs? What about the features? Should they stay at the board level or at the CPU level?

kYc0o commented:
Waiting for #9234.

kYc0o commented:
Now waiting for #9866.

kYc0o commented:
Not waiting anymore. Rebased to current master to solve conflicts.

I changed my mind about the features. The rule is that boards must provide their own features if they're actually accessible on board. 

Besides the arduino boards, which might share the same features as they have the same layout regarding hardware, I believe each atmega based board must provide its own list of provided features, since all of them are quite different. Therefore, some peripherals might not be available for all boards, etc.

The only board missing this is the jiminy-mega256rfr2, @Josar can you provide a list of the features that are actually availabe on board?

Josar commented:
@kYc0o should be the same as the arduinos. Other MCU features are accesible but not implemented jet.

Besides that the CPU ID is implemented in the CPU, it might be a feature which could be ported to common.
```
# Put defined MCU peripherals here (in alphabetical order)
FEATURES_PROVIDED += periph_adc
FEATURES_PROVIDED += periph_gpio periph_gpio_irq
FEATURES_PROVIDED += periph_i2c
FEATURES_PROVIDED += periph_spi
FEATURES_PROVIDED += periph_timer
FEATURES_PROVIDED += periph_uart

# Various other features (if any)

# The board MPU family (used for grouping by the CI system)
FEATURES_MCU_GROUP = avr6

# Include features provided by the CPU
include $(RIOTCPU)/atmega256rfr2/Makefile.features
```

smlng commented on makefiles/arch/atmega.inc.mk:
```diff
@@ -0,0 +1,39 @@
+endif
+
+# the atmel port uses uart_stdio
+USEMODULE += uart_stdio
```
typo: this is actually named `stdio_uart`

miri64 replied:
The module was actually renamed in https://github.com/RIOT-OS/RIOT/commit/d55616a7f56ca17440 (so after this PR was opened) so not sure this really is a typo ;-)

smlng commented on boards/mega-xplained/Makefile.include:
```diff
@@ -2,23 +2,14 @@
-export PORT_DARWIN ?= $(firstword $(sort $(wildcard /dev/tty.usbserial*)))
-export BAUD        ?= 9600
+PORT_LINUX  ?= /dev/ttyACM0
+PORT_DARWIN ?= $(firstword $(sort $(wildcard /dev/tty.usbmodem*)))
```
For an official Arduino its `usbserial` on macOS, I have a cheap-one where I've to set `PORT` explicitly anyway

smlng replied:
(forgot to say: have a Mac, too)

smlng commented:
to me this is good, please squash and rebase on latest master (had an issue which was solved in master while testing)

smlng commented:
@ZetaR60 and @aabadie are we good to go here?

miri64 commented:
I guess #7306 needs another rework if this is merged?

smlng commented:
> I guess #7306 needs another rework if this is merged?

might be, but rather minimal then - I think only `boards/arduino-leonardo/Makefile.include` would need some adaption, actually removing code (I think) ... but maybe @kYc0o can guide/help there

kYc0o commented:
Ping @aabadie @ZetaR60

kYc0o commented:
Weekly ping @aabadie @ZetaR60. I guess it didn't make it for the release?

smlng commented:
please rebase

ZetaR60 commented:
Sorry for the delay. The issues from my review seem to be resolved. Once it is rebased I will test on mega-xplained and arduino-uno.

maribu commented:
@kYc0o: ping

kYc0o commented:
Rebased.

maribu commented:
@ZetaR60: ping

(Sorry for making all the fuzz. I waiting for this PR to get merged in order to move two existing PRs forward and create a third on. Normally I'm less annoying ;-))

ZetaR60 commented:
Final double-check on mega-xplained looks good. Thanks for the cleanup @kYc0o !

kYc0o commented:
Thanks everyone for your review! :tada:
